oci/archive/reader.go

```diff
@@ -0,0 +1,197 @@
+	}
+
+	reader := Reader{
+		tempDirectory: dst,
+		path:          src,
+	}
+
+	if err := archive.NewDefaultArchiver().Untar(arch, dst, &archive.TarOptions{NoLchown: true}); err != nil {
+		return nil, perrors.Wrapf(err, "error untarring file %q", dst)
+	}
+
+	indexJSON, err := os.Open(filepath.Join(dst, "index.json"))
+	if err != nil {
+		return nil, err
+	}
+	defer indexJSON.Close()
+	reader.manifest = &imgspecv1.Index{}
+	if err := json.NewDecoder(indexJSON).Decode(reader.manifest); err != nil {
+		return nil, err
+	}
+
+	return &reader, nil
+}
+
+// NewReader returns a Reader for src. The caller should call Close() on the returned object
+func NewReaderForReference(ctx context.Context, sys *types.SystemContext, ref types.ImageReference) (*Reader, types.ImageReference, error) {
+	standalone, ok := ref.(ociArchiveReference)
+	if !ok {
+		return nil, nil, perrors.Errorf("Internal error: NewReader called for a non-oci/archive ImageReference %s", transports.ImageName(ref))
+	}
+	if standalone.archiveReader != nil {
+		return nil, nil, perrors.Errorf("Internal error: NewReader called for a reader-bound reference %s", standalone.StringWithinTransport())
+	}
+
+	reader, err := NewReader(ctx, sys, standalone.resolvedFile)
+	if err != nil {
+		return nil, nil, err
+	}
+	// src := standalone.resolvedFile
+	// arch, err := os.Open(src)
+	// if err != nil {
+	// 	return nil, err
+	// }
+	// defer arch.Close()
+
+	// dst, err := ioutil.TempDir(tmpdir.TemporaryDirectoryForBigFiles(sys), "oci")
+	// if err != nil {
+	// 	return nil, fmt.Errorf("error creating temp directory: %w", err)
+	// }
+
+	// reader := Reader{
+	// 	tempDirectory: dst,
+	// 	path:          src,
+	// }
+
+	succeeded := false
+	defer func() {
+		if !succeeded {
+			reader.Close()
+		}
+	}()
+	// if err := archive.NewDefaultArchiver().Untar(arch, dst, &archive.TarOptions{NoLchown: true}); err != nil {
+	// 	return nil, fmt.Errorf("error untarring file %q: %w", dst, err)
+	// }
+
+	// indexJSON, err := os.Open(filepath.Join(dst, "index.json"))
+	// if err != nil {
+	// 	return nil, err
+	// }
+	// defer indexJSON.Close()
+	// reader.manifest = &imgspecv1.Index{}
+	// if err := json.NewDecoder(indexJSON).Decode(reader.manifest); err != nil {
+	// 	return nil, err
+	// }
+	readerRef, err := newReference(standalone.resolvedFile, standalone.image, -1, reader, nil)
+	if err != nil {
+		return nil, nil, err
+	}
+	succeeded = true
+	return reader, readerRef, nil
+}
+
+// ListResult wraps the image reference and the manifest for loading
+type ListResult struct {
+	ImageRef           types.ImageReference
+	ManifestDescriptor imgspecv1.Descriptor
+}
+
+// List returns a slice of manifests included in the archive
+func (r *Reader) List() ([]ListResult, error) {
+	var res []ListResult
+
+	for _, md := range r.manifest.Manifests {
+		refName := internal.NameFromAnnotations(md.Annotations)
+		index := -1
+		if refName != "" {
+			index = 1
+		}
+		ref, err := newReference(r.path, refName, index, r, nil)
```

Comment:
This looks rather incorrect:
- It is setting `index` exactly in the case where it is not needed (and `newReference` would reject such an input)
- The index needs to actually point at the right manifest, not a hard-coded `1`.